### PR TITLE
Added NeoPixel FeatherWing

### PR DESCRIPTION
**Pull request by makermelissa**
Ok, I finished the NeoPixel FeatherWing, which is based on my DotStar FeatherWing work.

### CODE REVIEW

File: adafruit_featherwing/dotstar_featherwing.py
Comment (kattni) [inline adafruit_featherwing/dotstar_featherwing.py]: I would consider naming this `matrix` instead of `display` as we have a lot of display libraries within which we use that term. There are some FeatherWings with displays (or maybe only one?) and it might be a little confusing.

**Review by kattni — CHANGES_REQUESTED** [review]
This looks great! Thank you for extending the library the way you did, I really like that you didn't duplicate a lot of code. 

I have a suggestion that affects a number of lines, but I won't comment on each one as there are many.

**Comment by caternuson** [comment]
Interesting inheritance. How about refactoring everything into a third generic class that both DotStar and NeoPixel could derive from?

**Comment by makermelissa** [comment]
Thanks @kattni. Matrix sounds good. I was just trying to use a generic term. @caternuson I started going down that path but there realized there were less disables required for pylint this way. Either way would work though if you'd prefer I change it to that.

**Comment by makermelissa** [comment]
Added @kattni's suggested change

**Comment by caternuson** [comment]
@makermelissa What's an example of one of the pylint disables?

**Comment by makermelissa** [comment]
I can't remember the exact one, but I put all of the functionality in the base class and because it self._display (now matrix) hadn't been assigned until the extended classes were initialized, Pylint was complaining that the methods/properties of display that I was calling in the base class didn't exist.

**Comment by caternuson** [comment]
Did you have a stubbed out placeholder in the base class for `self._matrix`? I'm thinking just set it to `None` in the base class's `__int__()` and then set it to a DotStar/NeoPixel instance in the derived classes.

Something like `pixels` in `BaseClass` here:
```python
class DotStar:
    """DotStar class"""
    def __init__(self, clock, data):
        self.clock = clock
        self.data = data

    def show(self):
        """Show stuff."""
        print("I'm a DotStar with clock = {} and data = {}".format(self.clock, self.data))

class NeoPixel:
    """NeoPixel class"""
    def __init__(self, data):
        self.data = data

    def show(self):
        """Show stuff."""
        print("I'm a NeoPixel with data = {}".format(self.data))

class BaseThing:
    """Base class."""
    def __init__(self):
        self.pixels = None

    def show(self):
        """Show stuff."""
        self.pixels.show()

class DotStarThing(BaseThing):
    """DotStar class."""
    def __init__(self, clock, data):
        super().__init__()
        self.pixels = DotStar(clock, data)

class NeoPixelThing(BaseThing):
    """NeoPixel class."""
    def __init__(self, data):
        super().__init__()
        self.pixels = NeoPixel(data)
```

Test usage:
```python
Adafruit CircuitPython 3.1.2 on 2019-01-07; Adafruit ItsyBitsy M4 Express with samd51g19
>>> import class_test
>>> d = class_test.DotStarThing(1,2)
>>> n = class_test.NeoPixelThing(8)
>>> d.show()
I'm a DotStar with clock = 1 and data = 2
>>> n.show()
I'm a NeoPixel with data = 8
>>> 
```

**Comment by makermelissa** [comment]
I tried setting it to None and when I tried to set it in the child class I got pylint errors such as
```
E:117,25: Value 'self._display' is unsubscriptable (unsubscriptable-object)
E:119,16: 'self._display' does not support item assignment (unsupported-assignment-operation)
```

**Comment by caternuson** [comment]
Can you put the full code up in a branch in your repo and link it here.

**Comment by makermelissa** [comment]
You can see it in one of my above commits. Link is https://github.com/adafruit/Adafruit_CircuitPython_FeatherWing/pull/25/commits/b862de51b27ec0c948e4ce684a0a18f6221de339

**Comment by caternuson** [comment]
Hmmm. Quick scan of that, yah, that's what I'm thinking. Seems OK. Wonder if you're hitting this?
https://github.com/PyCQA/pylint/issues/1498

What if you ignore pylint and just try running it? Was it functional?

**Comment by makermelissa** [comment]
Yeah, it was functional. That's why I mentioned pylint disables.

**Comment by makermelissa** [comment]
And yes, that sounds like the issue I ran into.

**Comment by caternuson** [comment]
If you'd rather have gone with that class arrangement, I'd say it's OK to go ahead and add the disables. Seems like you only went with the current arrangement because of what appears to be a still open pylint bug.

You can add the disable globally, so you don't have to pepper it all over the place.

**Comment by makermelissa** [comment]
That's correct. I'll go ahead and make the changes and get them pushed.

**Comment by makermelissa** [comment]
Ok, it is now using the base class, has @kattni's suggestion, passes pylint, and functions perfectly.

**Comment by caternuson** [comment]
I'd suggest keeping all the methods public in the `PixelMatrixFeatherWing` class (no leading `_`). Then your two subclasses would only need to override `__init__` to plumb in an actual `_matrix` object.

Also, move `PixelMatrixFeatherWing` to a separate file and then import it in the other two.

**Comment by makermelissa** [comment]
Yeah, the main reason I did it with private methods was to be able to provide FeatherWing specific examples for each function for purposes of RTD. However, I can make is so the docstrings reflect one wing or the other.

**Comment by makermelissa** [comment]
I moved PixelMatrix to a separate file. However, I'm waiting on a response about suggestions for code examples and RTD. In other words, if I remove all the public function calls from NeoPixel and DotStar then it will show RTD generated comments from the docstring of PixelMatrix on both files. (i.e. If I use the DotStar example it will show up under NeoPixel wing or vice versa) I would think most people probably won't have both wings. Plus PixelMatrix was never intended to be used on it's own since it's a base class. I could remove the example, but I feel like it would be more helpful to the end user to have those.

**Comment by tannewt** [comment]
RTD comments on both sounds good to me. Why would you need to remove the example?

**Comment by makermelissa** [comment]
Unless there's a way of doing it that I'm not aware of, I believe RTD will split off whatever example is in the base class to the extended classes.

**Comment by tannewt** [comment]
I'd move the full function examples into external example files. Then, in the function just include the related code rather than the full example and refer to the full example.

**Comment by makermelissa** [comment]
Ok @tannewt, thanks. I'll go ahead and do that. I think including the examples in the comments seems to only work well if classes don't share code. I'm doing something similar with the Alphanum/7-segment displays too so I'll apply the same approach.

**Comment by tannewt** [comment]
Great! Happy to help unblock you. Thank you for all of the awesome changes.

**Comment by makermelissa** [comment]
I went ahead and removed all of the overrides except shift_up and shift_down in the neopixel featherwing library since they are laid out vertically in reverse of each other and shift_up on the neopixel being the same direction as shift_down on the dotstar.

**Comment by caternuson** [comment]
Nice! Looks good. I like it from a code layout / inheritance stand point.

You've tested this with hardware on both Wings?

**Comment by makermelissa** [comment]
Yes @caternuson, I tested all examples on both wings.

**Review by caternuson — APPROVED** [review]

**Comment by caternuson** [comment]
@kattni I approved for CP Librarians. You good with changes too?

**Review by kattni — APPROVED** [review]
Looks good! Thanks again!

**Pull request merged**
